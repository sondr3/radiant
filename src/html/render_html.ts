/**
 * This module provides functions for rendering HTML elements and documents to strings.
 * @module
 */

import { PrettyPrinter } from "../pretty_printer.js";
import type { HTMLElementAttributes } from "./html_attributes.js";
import type { HTMLDocument, HTMLElement, VoidHTMLElement } from "./html_element.js";
import type { HTMLTag, VoidHTMLTag } from "./html_tags.js";

/**
 * Renders an HTML element or void element to a string.
 *
 * @param tag - The HTML element or void element to render.
 * @param opts - HTML rendering options
 * @returns The rendered element as a string.
 * @throws Error if the element type is not supported.
 */
export const renderElement = <T extends HTMLTag, A extends HTMLElementAttributes, C>(
	tag: HTMLElement<T, A, C> | VoidHTMLElement<Extract<HTMLTag, VoidHTMLTag>, A>,
	opts?: { pretty: boolean },
): string => {
	return new PrettyPrinter(opts?.pretty ?? false, "html").printNode(tag);
};

/**
 * Renders an HTML document to a string.
 *
 * @param doc - The HTML document to render.
 * @param opts - HTML rendering options
 * @returns The rendered HTML document as a string.
 */
export const renderDocument = (doc: HTMLDocument, opts?: { pretty: boolean }): string => {
	return new PrettyPrinter(opts?.pretty ?? false, "html").print(doc);
};
