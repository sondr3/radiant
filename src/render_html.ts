/**
 * This module provides functions for rendering HTML elements and documents to strings.
 * @module
 */

import type { HTMLElementAttributes } from "./attributes.js";
import type { HTMLDocument, HTMLElement, VoidHTMLElement } from "./html_element.js";
import { PrettyPrinter } from "./pretty_printer.js";
import type { HTMLTag, VoidHTMLTag } from "./tags.js";

/**
 * Renders an HTML element or void element to a string.
 *
 * @param tag - The HTML element or void element to render.
 * @returns The rendered element as a string.
 * @throws Error if the element type is not supported.
 */
export const renderElement = <T extends HTMLTag, A extends HTMLElementAttributes, C>(
	tag: HTMLElement<T, A, C> | VoidHTMLElement<Extract<HTMLTag, VoidHTMLTag>, A>,
	opts?: { pretty: boolean },
): string => {
	return new PrettyPrinter(opts?.pretty ?? false, "html").printNode(tag);
};

/**
 * Renders an HTML document to a string.
 *
 * @param doc - The HTML document to render.
 * @returns The rendered HTML document as a string.
 */
export const renderDocument = (doc: HTMLDocument, opts?: { pretty: boolean }): string => {
	return new PrettyPrinter(opts?.pretty ?? false, "html").print(doc);
};
