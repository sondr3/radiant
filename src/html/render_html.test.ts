import { describe, test } from "vitest";
import { h } from "./index.js";
import { renderDocument, renderElement } from "./render_html.js";

describe("HTML rendering", () => {
	test("works without pretty", ({ expect }) => {
		const input = h.document(h.doctype(), h.html(h.head(h.title("Hello, world!"))));

		expect(renderDocument(input)).toMatchSnapshot();
	});

	test("works with pretty", ({ expect }) => {
		const doc = h.document(
			h.doctype(),
			h.html(
				h.head(h.meta({ charset: "utf-8" }), h.title("Hello, world!")),
				h.body(
					h.h1({ class: "blah" }, "Hello, world!"),
					h.p("This is a bunch of text... blah blah blah"),
					h.p("With some more ", h.b("bold"), " text."),
				),
			),
		);

		expect(renderDocument(doc, { pretty: true })).toMatchSnapshot();
	});

	test("works with whitespace sensitive tags", ({ expect }) => {
		const doc = h.body(
			h.p("This is some regular text..."),
			h.pre(`     And this is a very long line of text that should be preserved
        
  this has
some 
    weird whitespace`),
		);

		expect(renderElement(doc, { pretty: true })).toMatchSnapshot();
	});

	test("nesting pre and code works", ({ expect }) => {
		const elem = h.pre(h.code("\n  def some_code():\n    pass\n"));
		expect(renderElement(elem)).toMatchSnapshot();
	});

	test("it renders raw HTML correctly", ({ expect }) => {
		const elem = h.div(h.p("Hello, world!"), h.raw("<p>Raw <span>frickin'</span> HTML</p>"));
		expect(renderElement(elem)).toMatchSnapshot();
	});
});
