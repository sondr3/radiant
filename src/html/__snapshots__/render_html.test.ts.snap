// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HTML rendering > it renders raw HTML correctly 1`] = `"<div><p>Hello, world!</p><p>Raw <span>frickin'</span> HTML</p></div>"`;

exports[`HTML rendering > nesting pre and code works 1`] = `
"<pre><code>
  def some_code():
    pass
</code></pre>"
`;

exports[`HTML rendering > works with pretty 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Hello, world!</title>
  </head>
  <body>
    <h1 class="blah">Hello, world!</h1>
    <p>This is a bunch of text... blah blah blah</p>
    <p>With some more <b>bold</b> text.</p>
  </body>
</html>"
`;

exports[`HTML rendering > works with whitespace sensitive tags 1`] = `
"<body>
  <p>This is some regular text...</p>
  <pre>     And this is a very long line of text that should be preserved
        
  this has
some 
    weird whitespace</pre>
</body>"
`;

exports[`HTML rendering > works without pretty 1`] = `"<!DOCTYPE html><html><head><title>Hello, world!</title></head></html>"`;
